ostd/src/mm/heap_allocator/mod.rs

```diff
@@ -54,25 +55,50 @@ impl LockedHeapWithRescue {
         self.heap
             .call_once(|| SpinLock::new(Heap::new(start as usize, size)));
     }
+
+    /// SAFETY: The range [start, start + size) must be a valid memory region.
+    unsafe fn add_to_heap(&self, start: usize, size: usize) {
+        self.heap
+            .get()
+            .unwrap()
+            .disable_irq()
+            .lock()
+            .add_memory(start, size);
+    }
+
+    fn avoid_low_memory(&self, remain_bytes: usize, layout: Layout) {
+        if remain_bytes <= PAGE_SIZE * 4 {
+            debug!(
+                "Low memory in heap allocator, try to call rescue. Remaining bytes: {:x?}",
+                remain_bytes
+            );
+            // We don't care if the rescue returns ok or not since we can still do heap allocation.
+            let _ = rescue(self, &layout);
+        }
+    }
 }
 
 unsafe impl GlobalAlloc for LockedHeapWithRescue {
     unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
-        let mut heap = self.heap.get().unwrap().disable_irq().lock();
+        let _guard = disable_local();
 
-        if let Ok(allocation) = heap.allocate(layout) {
-            return allocation as *mut u8;
+        let res = self.heap.get().unwrap().lock().allocate(layout);
+        if let Ok((allocation, remain_bytes)) = res {
+            self.avoid_low_memory(remain_bytes, layout);
+            return allocation;
         }
 
-        // Avoid locking self.heap when calling rescue.
-        if rescue(&mut heap, &layout).is_err() {
+        if rescue(self, &layout).is_err() {
             return core::ptr::null_mut::<u8>();
         }
 
-        heap.allocate(layout)
-            .map_or(core::ptr::null_mut::<u8>(), |allocation| {
-                allocation as *mut u8
-            })
+        let res = self.heap.get().unwrap().lock().allocate(layout);
+        if let Ok((allocation, remain_bytes)) = res {
+            self.avoid_low_memory(remain_bytes, layout);
+            allocation
+        } else {
+            core::ptr::null_mut::<u8>()
```

Comment:
When the heap allocation fails, we can have even more reasons to rescue right?

Reply:
When the heap allocation fails, the frame allocator does not have enough memory for the heap allocator. Calling the rescue here does not help with the situation.

Reply:
Ok. I overlooked the call to `self.rescue()` before the allocation that returns null.